OPHD/States/MainMenuState.cpp

```diff
@@ -148,37 +148,63 @@ void MainMenuState::enableButtons()
  */
 void MainMenuState::fileIoAction(const std::string& filePath, FileIo::FileOperation fileOp)
 {
-	if (fileOp != FileIo::FileOperation::FILE_LOAD) { return; }
-	if (filePath.empty()) { return; }
-
-	std::string filename = constants::SAVE_GAME_PATH + filePath + ".xml";
-
-	if (!Utility<Filesystem>::get().exists(filename))
+	if (fileOp == FileIo::FileOperation::FILE_SAVE)
 	{
-		doNonFatalErrorMessage("Load Failed", "File '" + filename + "' was not found.");
 		return;
 	}
 
-	try
+	if (filePath.empty())
 	{
-		checkSavegameVersion(filename);
-
-		GameState* gameState = new GameState();
-		MapViewState* mapview = new MapViewState(gameState->getMainReportsState(), filename);
-		mapview->_initialize();
-		mapview->activate();
-
-		gameState->mapviewstate(mapview);
-		mReturnState = gameState;
+		return;
+	}
 
-		Utility<Renderer>::get().fadeOut(constants::FADE_SPEED);
-		Utility<Mixer>::get().fadeOutMusic(constants::FADE_SPEED);
+	if (fileOp == FileIo::FileOperation::FILE_LOAD)
+	{
+		std::string filename = constants::SAVE_GAME_PATH + filePath + ".xml";
+
+		if (!Utility<Filesystem>::get().exists(filename))
+		{
+			doNonFatalErrorMessage("Load Failed", "File '" + filename + "' was not found.");
+			return;
+		}
+
+		try
+		{
+			checkSavegameVersion(filename);
+
+			GameState* gameState = new GameState();
+			MapViewState* mapview = new MapViewState(gameState->getMainReportsState(), filename);
+			mapview->_initialize();
+			mapview->activate();
+
+			gameState->mapviewstate(mapview);
+			mReturnState = gameState;
+
+			Utility<Renderer>::get().fadeOut(constants::FADE_SPEED);
+			Utility<Mixer>::get().fadeOutMusic(constants::FADE_SPEED);
+		}
+		catch (const std::exception& e)
+		{
+			mReturnState = this;
+			doNonFatalErrorMessage("Load Failed", e.what());
+		}
 	}
-	catch (const std::exception& e)
+
+	if (fileOp == FileIo::FileOperation::FILE_DELETE)
 	{
-		mReturnState = this;
-		doNonFatalErrorMessage("Load Failed", e.what());
+		std::string filename = constants::SAVE_GAME_PATH + filePath + ".xml";
+
+		if (!Utility<Filesystem>::get().exists(filename))
+		{
+			doNonFatalErrorMessage("Delete Failed", "File '" + filename + "' was not found.");
+			return;
+		}
+		else
+		{
+			Utility<Filesystem>::get().del(filename);
+		}
```

Comment:
Looks like the delete code is duplicated. I think this is a further sign that perhaps the `FileIo` dialog should handle this responsibility.

Reply:
I based the design of this on the Windows API -- the FileIO dialog's are a standard function but the dialog itself doesn't do anything, it returns to the caller the action that the user wants to do. It's then the responsibility of the caller to decide how to handle it.

I think you're right when you state above that the Filesystem will throw if there's a problem so we should just let the Filesystem exception code do the work and use a try/catch block. So the above highlighted code block could look like this:

```cpp
if (fileOp == FileIo::FileOperation::FILE_DELETE)
{
	std::string filename = constants::SAVE_GAME_PATH + filePath + ".xml";

	try
	{
		Utility<Filesystem>::get().del(filename);
	}
	catch(const std::exception& e)
	{
		doNonFatalErrorMessage("Delete Failed", e.what());
	}
}
```

I assume this was based off the previously existing code which is a little on the old side and was developed long before the current state of the Filesystem code so I think it makes sense to start the improvement with a new block of code.

Note that the try/catch block can be omitted thought I opt to catch here because these are exceptions that can almost universally be recovered from in this context (file operations).